lib/ecto/adapters/poolboy/worker.ex

```diff
@@ -0,0 +1,188 @@
+  alias Ecto.Adapters.Pool
+
+  @type modconn :: {module :: atom, conn :: pid}
+
+  @spec start_link({module, Keyword.t}) :: {:ok, pid}
+  def start_link({module, params}) do
+    GenServer.start_link(__MODULE__, {module, params}, [debug: [:log]])
+  end
+
+  @spec checkout(pid, timeout) :: {Pool.mode, modconn} | {:error, Exception.t}
+  def checkout(worker, timeout) do
+    GenServer.call(worker, :checkout, timeout)
+  end
+
+  @spec open_transaction(pid, timeout) ::
+    {Pool.mode, modconn} | {:error, Exception.t}
+  def open_transaction(worker, timeout) do
+    GenServer.call(worker, :open_transaction, timeout)
+  end
+
+  @spec close_transaction(pid) :: :ok
+  def close_transaction(worker) do
+    GenServer.cast(worker, :close_transaction)
+  end
+
+  @spec break(pid, timeout) :: :ok
+  def break(worker, timeout) do
+    GenServer.call(worker, :break, timeout)
+  end
+
+  @spec transaction_mode(pid, :raw | :sandbox, timeout) ::
+    :ok | {:error, :noconnect}
+  def transaction_mode(worker, mode, timeout) do
+    GenServer.call(worker, {:transaction_mode, mode}, timeout)
+  end
+
+  ## Callbacks
+
+  def init({module, params}) do
+    Process.flag(:trap_exit, true)
+    lazy? = Keyword.get(params, :lazy, true)
+
+    unless lazy? do
+      case module.connect(params) do
+        {:ok, conn} ->
+          conn = conn
+        _ ->
+          :ok
+      end
+    end
+
+    {:ok, %{conn: conn, params: params, transaction: nil, mode: :raw,
+            module: module}}
+  end
+
+  ## Break
+
+  def handle_call(:break, _from, %{mode: :sandbox} = s) do
+    {:reply, :ok, demonitor(s)}
+  end
+
+  def handle_call(:break, _from, s) do
+    s = s
+      |> demonitor()
+      |> disconnect()
+    {:reply, :ok, s}
+  end
+
+  ## Mode change
+
+  def handle_call({:transaction_mode, _}, _from, %{conn: nil} = s) do
+    {:reply, {:error, :noconnect}, s}
+  end
+  def handle_call({:transaction_mode, :raw}, _from, %{mode: :sandbox} = s) do
+    {:reply, :ok, %{s | mode: :raw}}
+  end
+  def handle_call({:transaction_mode, :sandbox}, _from, %{mode: :raw} = s) do
+    {:reply, :ok, %{s | mode: :sandbox}}
+  end
+  def handle_call({:transaction_mode, mode}, _from, %{mode: mode} = s) do
+    {:stop, :already_mode, s}
+  end
+
+  ## Lazy connection handling
+
+  def handle_call(request, from, %{conn: nil, params: params, module: module} = s) do
+    case module.connect(params) do
+      {:ok, conn}   -> handle_call(request, from, %{s | conn: conn})
+      {:error, err} -> {:reply, {:error, err}, s}
+    end
+  end
+
+  ## Checkout
+
+  def handle_call(:checkout, _, %{mode: mode} = s) do
+    {:reply, {mode, modconn(s)}, s}
+  end
+
+  ## Open transaction
+
+  def handle_call(:open_transaction, {pid, _},
+  %{transaction: nil, mode: mode} = s) do
+    {:reply, {mode, modconn(s)}, monitor(pid, s)}
+  end
+
+  def handle_call(:open_transaction, from,
+  %{transaction: {client, _}, mode: mode} = s) do
+    if Process.is_alive?(client) or mode === :sandbox do
```

Comment:
Is this chain of events possible
* Client crashed and starts exiting
* client :DOWN to poolboy
* poolboy receives :DOWN and sends worker pid to new client
* new client sends open_transaction/checkout
* old client :DOWN to worker

Maybe this is one to ask Concuerror.